kb/how-to/rewrite-branch-for-coding-style.md

```diff
@@ -0,0 +1,163 @@
+# Switching to the new code style
+
+In January 2023, Mbed TLS switched to a new coding style, which is now enforced by continuous integration tests. This document explains how to adapt ongoing work. Please read this document if you started work on a branch of Mbed TLS that has the old coding style, and you want to contribute this work to Mbed TLS or you are already doing so.
+
+This concerns both maintained branches of Mbed TLS: `development` and `mbedtls-2.28`.
+
+## Coding style overview
+
+The following table summarizes the old and new coding style (before/after January 2023).
+
+| Aspect | Old | New |
+| ------ | --- | --- |
+| Indentation | 4 spaces | 4 spaces |
+| Indentation of `case` | one level deeper than `switch` | one level deeper than `switch` |
+| Maximum line length | 80 preferred | 80 preferred |
+| **Spacing inside parentheses** | always except `define` and casts: `if( f( x ) )` | **never**: `if (f(x))` |
+| Spacing inside square brackets | no: `a[i]` | no: `a[i]` |
+| **Spacing before opening parenthesis** | no: `if( f( x ) )` | **after keyword or binary operator**: `if (f(x))` |
+| Spacing before opening square bracket | no: `a[i]` | no: `a[i]` |
+| Spacing after unary prefix operator | optional: `!c` or `! c` | **no**: `!c` |
+| Spacing around binary operator | yes: `( x + y ) * z` | yes: `(x + y) * z` |
+| Parentheses around return value | mandatory: `return( x );` | **generally not**: `return x;` |
+| **Single-line if/loop body** | separate line, braces optional | separate line, **braces mandatory** |
+| Opening brace | on its own line | **on the same line** except for functions |
+
+For a more detailed description of the coding style, please refer to the [coding standards](../development/mbedtls-coding-standards.md).
+
+## Migrating existing work: the theory
+
+If you started working on a branch of Mbed TLS with the old code style, any part of the code that has been both modified in your branch and restyled in the main branch will lead to a merge conflict. To avoid that, you must migrate your work to the new style.
+
+It is not enough to rewrite the final state of your work to the new format, because that will still be a merge conflict (between just restyling, and semantically meaningful work followed by restyling). You must rebase-and-restyle your branch. This document explains how and when.
+
+The Mbed TLS project provides a script [`scripts/code_style_rewrite_branch.py`](https://github.com/Mbed-TLS/mbedtls/blob/development/scripts/code_style_rewrite_branch.py) which takes a Git branch with work done in the old code style, and migrates it to the new code style. Suppose your work is in a branch `mywork`, and it's intended to be submitted to some target branch (such as `development`). The migration works in two phases:
```

Comment:
Suggestion: regarding the details of how the script works, should we first say something like "in most cases you can skip this and just use the script, but the details are useful if something goes wrong and you need to fix something manually"?

I think the explanations are great but I don't want to scare users when the process is expected to be pretty simple most of the time.

Perhaps we should even have the details in a separate section at the end? And here just introduce the script as a black box: what are its inputs and the result. (So, perhaps just the initial and final git diagram - I do think the diagrams are helpful.)

Reply:
I agree: this document is currently structured from an architectural point of view, and I want to reorder it to be a proper user document.